### PR TITLE
Bump ckb version to v0.100.0 and adapt hardfork

### PR DESCRIPTION
**Pull request by classicalliu**

### CODE REVIEW

**Review by Flouse — DISMISSED** [review]

File: crates/common/src/lib.rs [hunk 1]
```diff
@@ -24,3 +24,11 @@ cfg_if::cfg_if! {
         use alloc::vec;
     }
 }
+
+lazy_static::lazy_static! {
```
Comment (jjyr) [inline crates/common/src/lib.rs:28]: The `gw-common` is supposed used by both godwoken node & scripts. It should only depends on `no-std` crates.

We should create a new crate `gw-ckb-hardfork` and move these code to there.

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/common/Cargo.toml [hunk 2]
```diff
@@ -12,6 +12,9 @@ sparse-merkle-tree = { version = "0.5.2-rc1", default-features = false }
 merkle-cbt = { version = "0.3.0", default-features = false }
 gw-hash = { path = "../hash" }
 thiserror = { version = "1.0", optional = true }
+lazy_static = "1.4"
```
Comment (jjyr) [inline crates/common/Cargo.toml:15]: Pls remove these crates, see the comment below.

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/common/Cargo.toml (hunk 2 above)
Comment (classicalliu) [reply crates/common/Cargo.toml]: updated

File: crates/generator/src/generator.rs [hunk 3]
```diff
@@ -44,6 +45,27 @@ use ckb_vm::{
     DefaultMachineBuilder, SupportMachine,
 };
 
+struct AsmCoreMachineParams {
+    pub vm_isa: u8,
+    pub vm_version: u32,
+}
+
+impl AsmCoreMachineParams {
+    pub fn new(vm_version: u32) -> AsmCoreMachineParams {
```
Comment (jjyr) [inline crates/generator/src/generator.rs:54]:
```suggestion
    pub fn with_version(vm_version: u32) -> Result<AsmCoreMachineParams> {
```

File: crates/generator/src/generator.rs [hunk 4]
```diff
@@ -44,6 +45,27 @@ use ckb_vm::{
...
+        if vm_version == 0 {
+            AsmCoreMachineParams {
+                vm_isa: ckb_vm::ISA_IMC,
+                vm_version: ckb_vm::machine::VERSION0,
+            }
+        } else {
```
Comment (jjyr) [inline crates/generator/src/generator.rs:60]:
```suggestion
        } else if vm_version == 1 {
```

File: crates/generator/src/generator.rs [hunk 5]
```diff
@@ -44,6 +45,27 @@ use ckb_vm::{
...
+            AsmCoreMachineParams {
+                vm_isa: ckb_vm::ISA_IMC | ckb_vm::ISA_B | ckb_vm::ISA_MOP,
+                vm_version: ckb_vm::machine::VERSION1,
+            }
+        }
```
Comment (jjyr) [inline crates/generator/src/generator.rs:65]: return error if vm_version is invalid

File: crates/block-producer/src/runner.rs [hunk 6]
```diff
@@ -159,6 +160,23 @@ async fn poll_loop(
             // update tip
             tip_number = raw_header.number().unpack();
             tip_hash = block.header().hash().into();
+
+            // update global hardfork info
+            let hardfork_switch = rpc_client.get_hardfork_switch().await?;
+            let rpc32_epoch_number = hardfork_switch.rfc_0032();
+            *GLOBAL_HARDFORK_SWITCH.lock().await = hardfork_switch;
```
Comment (jjyr) [inline crates/block-producer/src/runner.rs:167]: Skip the updating if value isn't changed

File: crates/block-producer/src/runner.rs [hunk 7]
```diff
@@ -159,6 +160,23 @@ async fn poll_loop(
...
+
+            // update global current epoch number
+            let current_epoch_number = rpc_client.get_current_epoch_number().await?;
+            *GLOBAL_CURRENT_EPOCH_NUMBER.lock().await = current_epoch_number;
```
Comment (jjyr) [inline crates/block-producer/src/runner.rs:171]:
```suggestion
            let mut global_epoch_number = GLOBAL_CURRENT_EPOCH_NUMBER.lock().await;
            if *global_epoch_number != current_epoch_number {
              *global_epoch_number = current_epoch_number;
            }
```

File: crates/block-producer/src/runner.rs [hunk 8]
```diff
@@ -159,6 +160,23 @@ async fn poll_loop(
...
+
+            // update global vm version
+            let vm_version: u32 = if current_epoch_number >= rpc32_epoch_number {
+                1
+            } else {
+                0
+            };
+            *GLOBAL_VM_VERSION.lock().await = vm_version;
```
Comment (jjyr) [inline crates/block-producer/src/runner.rs:179]:
```suggestion
            let mut global_vm_version = *GLOBAL_VM_VERSION.lock().await;
            if *global_vm_version != vm_version {
              *global_vm_version = vm_version;
            }
```

**Review by jjyr — CHANGES_REQUESTED** [review]

File: crates/generator/src/generator.rs (hunk 3 above)
Comment (classicalliu) [reply crates/generator/src/generator.rs]: updated to `Result<AsmCoreMachineParams, VMError>`

File: crates/generator/src/generator.rs (hunk 4 above)
Comment (classicalliu) [reply crates/generator/src/generator.rs]: updated

File: crates/generator/src/generator.rs (hunk 5 above)
Comment (classicalliu) [reply crates/generator/src/generator.rs:67]: updated

File: crates/block-producer/src/runner.rs (hunk 6 above)
Comment (classicalliu) [reply crates/block-producer/src/runner.rs]: updated

File: crates/block-producer/src/runner.rs (hunk 7 above)
Comment (classicalliu) [reply crates/block-producer/src/runner.rs]: updated

File: crates/block-producer/src/runner.rs (hunk 8 above)
Comment (classicalliu) [reply crates/block-producer/src/runner.rs]: updated

**Review by jjyr — APPROVED** [review]

**Pull request merged**